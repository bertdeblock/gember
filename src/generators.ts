import { generate } from "./generate.js";

export function generateComponent(
  name: string,
  packagePath: string,
  {
    classBased = false,
    nested,
    path,
    typescript = false,
  }: {
    classBased?: boolean;
    nested?: boolean;
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate({
    customTargetPath: path,
    entityName: name,
    generatorName: "component",
    nested,
    packagePath,
    targetDir: "components",
    templateFilename:
      (classBased ? "class-based" : "template-only") +
      (typescript ? ".gts" : ".gjs"),
  });
}

export function generateHelper(
  name: string,
  packagePath: string,
  {
    classBased = false,
    path,
    typescript = false,
  }: {
    classBased?: boolean;
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate({
    customTargetPath: path,
    entityName: name,
    generatorName: "helper",
    packagePath,
    targetDir: "helpers",
    templateFilename:
      (classBased ? "class-based" : "function-based") +
      (typescript ? ".ts" : ".js"),
  });
}

export function generateModifier(
  name: string,
  packagePath: string,
  {
    classBased = false,
    path,
    typescript = false,
  }: {
    classBased?: boolean;
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate({
    customTargetPath: path,
    entityName: name,
    generatorName: "modifier",
    packagePath,
    targetDir: "modifiers",
    templateFilename:
      (classBased ? "class-based" : "function-based") +
      (typescript ? ".ts" : ".js"),
  });
}

export function generateService(
  name: string,
  packagePath: string,
  {
    path,
    typescript = false,
  }: {
    path?: string;
    typescript?: boolean;
  } = {},
): Promise<void> {
  return generate({
    customTargetPath: path,
    entityName: name,
    generatorName: "service",
    packagePath,
    targetDir: "services",
    templateFilename: typescript ? "service.ts" : "service.js",
  });
}
