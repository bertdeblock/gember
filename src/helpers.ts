import type { EmberPackageJson } from "./types.js";

export function isV1Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 1 || version === undefined;
  }

  return false;
}

export function isV2Addon(packageJson: EmberPackageJson): boolean {
  if (isAddon(packageJson)) {
    const { version } = packageJson["ember-addon"] ?? {};

    return version === 2;
  }

  return false;
}

function isAddon(packageJson: EmberPackageJson): boolean {
  if (Array.isArray(packageJson.keywords)) {
    return packageJson.keywords.includes("ember-addon");
  }

  return false;
}
