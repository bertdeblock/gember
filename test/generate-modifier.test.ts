import fsExtra from "fs-extra";
import { readFile } from "node:fs/promises";
import { join } from "node:path";
import { afterEach, it } from "vitest";
import { generateModifier } from "../src/generators.ts";
import { copyBlueprint } from "./helpers.ts";

let cwd: string;

afterEach(() => fsExtra.remove(cwd));

it("generates a function-based `.js` modifier", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateModifier("foo", { cwd });

  const content = await readFile(join(cwd, "src/modifiers/foo.js"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a class-based `.js` modifier", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateModifier("foo", { classBased: true, cwd });

  const content = await readFile(join(cwd, "src/modifiers/foo.js"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a function-based `.js` modifier at a custom path", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateModifier("foo", { cwd, path: "src/-private" });

  const content = await readFile(join(cwd, "src/-private/foo.js"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a function-based `.ts` modifier", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateModifier("foo", { cwd, typescript: true });

  const content = await readFile(join(cwd, "src/modifiers/foo.ts"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a class-based `.ts` modifier", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateModifier("foo", { classBased: true, cwd, typescript: true });

  const content = await readFile(join(cwd, "src/modifiers/foo.ts"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});

it("generates a function-based `.ts` modifier at a custom path", async (ctx) => {
  cwd = await copyBlueprint("v2-addon");

  await generateModifier("foo", { cwd, path: "src/-private", typescript: true });

  const content = await readFile(join(cwd, "src/-private/foo.ts"), "utf-8");

  ctx.expect(content).toMatchSnapshot();
});
