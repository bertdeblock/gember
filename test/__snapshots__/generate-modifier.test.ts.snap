// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`generates a class-based \`.js\` modifier 1`] = `
"import Modifier from "ember-modifier";

export default class Foo extends Modifier {
  modify(element, positional, named) {}
}
"
`;

exports[`generates a class-based \`.ts\` modifier 1`] = `
"import Modifier from "ember-modifier";

type Named = {};
type Positional = [];
type Element = null;

export interface FooSignature {
  Args: {
    Named: Named;
    Positional: Positional;
  };
  Element: Element;
}

export default class Foo extends Modifier<FooSignature> {
  modify(element: Element, positional: Positional, named: Named) {}
}
"
`;

exports[`generates a function-based \`.js\` modifier 1`] = `
"import { modifier } from "ember-modifier";

export default modifier(function foo(element, positional, named) {});
"
`;

exports[`generates a function-based \`.js\` modifier at a custom path 1`] = `
"import { modifier } from "ember-modifier";

export default modifier(function foo(element, positional, named) {});
"
`;

exports[`generates a function-based \`.ts\` modifier 1`] = `
"import { modifier } from "ember-modifier";

export interface FooSignature {
  Args: {
    Named: {};
    Positional: [];
  };
  Element: null;
}

export default modifier<FooSignature>(function foo(element, positional, named) {});
"
`;

exports[`generates a function-based \`.ts\` modifier at a custom path 1`] = `
"import { modifier } from "ember-modifier";

export interface FooSignature {
  Args: {
    Named: {};
    Positional: [];
  };
  Element: null;
}

export default modifier<FooSignature>(function foo(element, positional, named) {});
"
`;
